# Results

## The_fcoex_ method 

The_fcoex_ tool was built from first principles to provide better understandability. Our first goal was to develop a smaller set of genes that globally captured the cellular diversity of a dataset.
For that, we decided to explore feature selection by symmetrical uncertainty, the correlation metric of FCBF, a popular feature selection algorithm for machine learning with little previous use in biomedical sciences . Symmetrical uncertainty relies on entropy (in the information-theory sense), which relies on categories for calculation. Thus, we implemented a set of heuristics to binarize gene expressions which can be accessed via the_fcoex_::discretize() function.

![
**Reclustering based on co-expression modules sheds light on cell diversity.**
The pre-labeled pbmc3k dataset, with original, SeuratData labels. B)_fcoex_ coexpression modules capture patterns across PBMC populations. C-D) Module-based reclustering  shows groupings of original clusters, visible both via UMAP projection and cluster x cluster heatmap. E) Upper-level, multilayer classification of the pbmc3k dataset based on_fcoex_ modules.
 ](images/fcoex_overview.png){#fig:overview height=7in .white}

As mutual information is a supervised method,_fcoex_ also needs pre-made cluster assignments obtained after running a standard scRNA-seq clustering pipeline (Fig. 1A). Cluster assignments convey information about the relations between cells and help to guide feature selection.

Fcoex, then, selects genes global markers, which might be specific to 1, 2, or more clusters; the common factor is that they provide information to tell clusters apart.
To find the coexpression-module, we inverted the FCBF redundancy removal algorithm as a heuristic to find redundant (co-expressed) gene expression patterns. (see Supplementary Methods for details).

The gene coexpression modules yielded by the pipeline are small by design (10s of genes per module), so to facilitate manual exploration of the coexpression landscape. Each module has one “header” gene, which expression pattern is most representative of the genes in the module.

The ultimate goal of the_fcoex_ pipeline is not necessarily the modules but to find biologically relevant populations._fcoex_ treats each module as a gene set to find cell populations, using only their expression the module to re-classify the cells. The new classifications offer a platform for the biomedical research to gather new insights on the dataset, as exemplified in Figure 1. 

## _fcoex_ recovers multi-hierarchy of blood types

To validate the_fcoex_ pipeline, we selected the well-known pbmc3k dataset from SeuratData, which contains around 2700 peripheral blood mononuclear cells (PBMC) with previously-defined cluster labels.

The standard_fcoex_ pipeline detected nine modules that capture different parts of the cellular diversity in the dataset. For example, module M8, containing cytotoxicity genes as PRF1 and GZMA, split the dataset into cytotoxic (NK and CD8) and non-cytotoxic cells. M2 (CD3D) split the dataset clearly in T-cells and non-T-cells. M5 (HLA-DRB1) grouped monocytes, B cells, and dendritic cells, all known antigen-presenting cells (APC). (Fig. 1D-E)

In general,_fcoex_ clusters combined biologically similar cell types of the original dataset. The clusterings, then, help to explore and classify upper cell classes by function (Fig. 1E). Even in that super well-studied dataset,_fcoex_ provided a new light on the shared functionality of some NK cells and macrophages: they both markedly express the CD16-coding gene FCGR3A, whose product is a key player in Antibody-dependent cellular cytotoxicity (ADCC). Thus, a complete functional classification of cells might want to include an ’ADCC-performing cells" class.