# Discussion 

Here we presented fcoex, a ready-to-use R/Bioconductor package for co-expression-based reclustering of single-cell RNA-seq data.
We note that other methods are increasingly available for co-expression analysis of single cells. The monocle R package [@wikidata:Q41861103], widely used for pseudotime analysis, has implemented algorithms for detecting co-expression modules (<https://cole-trapnell-lab.github.io/monocle3/docs/differential/#gene-modules>), and WGCNA, widely used in bulk transcriptomics, has also been applied to scRNA-seq [@wikidata:Q21284194] [@wikidata:Q62485122]. 

In principle, any of those algorithms could be used as input for our framework (and we provide code showing how to integrate them to `fcoex`). We note, though, that fcoex modules are generally smaller and provide module header genes, making it a sensible first-pass approach to explore the  multi-layered diversity in single-cell transcriptomics datasets. 
In that way, fcoex offers ways to explore data-driven classifications of cells, aligning itself with the challenges of the Human Cell Atlas and, specifically, of building ontologies of cell types in the single-cell era.



