# Discussion 

Here we report a pipeline to explore cell type divesity via ranked, lean coexpression modules. 

The multilayered view on gene expression is the core conttribution of fcoex to the arsenal of tools. 

We note that other methods are increasingly available for co-expression analysis of single cells. 

The popular frameworks Scanpy(https://scanpy.readthedocs.io/en/stable/tutorials.html#, https://github.com/theislab/scanpy/issues/72 ), Seurat(https://satijalab.org/seurat/) and OSCA (from Bioconductor, https://bioconductor.org/books/release/OSCA/#) do not include specific instructions on using co-expressions works or providing multi-layer clustering, but the monocle R package (https://www.nature.com/articles/nbt.2859), widely used for pseudotime analysis, has implemented module-based analysis for its version 3, using the Louvain community detection algorithm to get modules from UMAP dimensions.(https://cole-trapnell-lab.github.io/monocle3/docs/differential/#gene-modules).


A common option is to employ the WGCNA method [@wikidata:Q21284194] or its adaptations [@wikidata:Q62485122] to identify co-expression modules. 

In principle, any of those modules could be used as input of multilayered clustering, and are alternatives for the first step of the fcoex pipeline. 
We note, though, that in contrast to unranked, large modules found by previous pipelines , fcoex modules makes it easy for domain experts to take advantage of the data. 
For example, in this work, a master's student could pinpoint the 4 genes of the apela/Toddler pathway floating on an 18-gene gene set.
He would surely have had more trouble finding a biologically meaningful association in a set with hundreds of genes.

Fcoex streamlines the path from dataset to coexpression modules and to multi layered populations.

The "cell type forests" obtained with the parallel reclustering of cells sheds light at higher orders of classifications. 

The approach supported our "rediscovery" of B-cells as both a cell type on their own, but also as a subclass of Antigen Presenting Cells.

In that way, fcoex offers ways to explore data-driven classifications of cells, aligning itself with the challenges of the Human Cell Atlas and, specifically, of building ontologies of cell types in the single-cell era. (ADD REFERENCES)
